### DISCUSSION TITLE
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: session

### DISCUSSION THREAD

**Discussion by iamhimanshu0** [apache/airflow#22014] [General]
### Apache Airflow version

2.2.4 (latest released)

### What happened

I've installed apache-airflow version 2.2.4 on my system ubuntu 20.0.4 LTS and i'm using these steps
- export AIRFLOW_HOME=~/airflow
- pip3 install apache-airflow
- airflow db init
- airflow users create --username admin --firstname XXXX --lastname XXX --role Admin --email XXX@gmail.com

When I'm running my server using this command 
- airflow webserver -p 8080
I've got an error saying 
`sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: session`

![Screenshot from 2022-03-05 10-27-20](https://user-images.githubusercontent.com/23204620/156868669-c7e41620-76a4-4c8e-a806-993cfed580af.png)

Can anyone help me with what I'm doing wrong? 
Thanks

### Operating System

Ubuntu 20.0.4

### Deployment

Virtualenv installation

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Please (as instructed when opening an issue) refrain from opening Airlflow issues where you have problems with your configuration.
Open Discussion instead!

**Reply by potiuk** [reply]
You should make sure AIRFLOW_HOME and all other configuration is the same for all your components. You likely did not have the same AIRFLOW_HOME set when you run `airlfow webserver`

**Reply by iamhimanshu0** [reply]
I try with this also like in VsCode I did
- export AIRFLOW_HOME=.
Then run the ```airflow db init``` and ```airflow webserver```
but still the same problem?

**Reply by potiuk** [reply]
Make sure you have everything set the same. since you have VSCode, debug it to see if for sure you are using the same sqlalchemy DB.

**Reply by iamhimanshu0** [reply]
yes, I did check everything is fine sqlalchemy DB, and all the airflow services but, still getting the error?

**Reply by potiuk** [reply]
You have some configuration wrong. Can't help you more than that. For some reason when you run whatever you run different DB is used. 

BTW. I am not sure what you are doing but in case you run unit tests, you know you should run them with `pytest` and you know that when  you are running tests, the unit test mode kicks in and  you know that there is a different database used  and you can reset it it `--with-db-init` switch of Pytest?

https://github.com/apache/airflow/blob/main/TESTING.rst#airflow-unit-tests

**Reply by iamhimanshu0** [reply]
I did try everything u said but sadly got the same results? any help?

Does it seem like there is something wrong with the session_id?

```[SQL: SELECT session.id AS session_id_1, session.session_id AS session_session_id, session.data AS session_data, session.expiry AS session_expiry 
FROM session 
WHERE session.session_id = ?
 LIMIT ? OFFSET ?]
[parameters: ('347fde8a-01bc-4074-9f3c-87f1dc7ec4de', 1, 0)]
(Background on this error at: http://sqlalche.me/e/13/e3q8)```

**Reply by potiuk** [reply]
No. You miss session table. You have an empty database most likely,

**Reply by iamhimanshu0** [reply]
but I did ```airflow db init``` and it's working fine, I don't know what went wrong...

**Reply by potiuk** [reply]
I do not know either

**Reply by iamhimanshu0** [reply]
so can we move this discussion to the issues section? cause I've tried everything but still not working?

**Reply by potiuk** [reply]
It does not give you anything if we do.what do you think would happen if we do?.


There is no reproducible csse to be able to diagnose it. Unless you help with making a reproducible case, there is no way anyone can help, i am afraid.

**Reply by nareshgiduthuri1** [reply]
Hi,

I was new to airflow and i was also facing same issue. Followed the same pattern of installation. Not sure how to resolve this...Kindly anyone please help

**Reply by potiuk** [reply]
@nareshgiduthuri1  - could you please describe what exactly you have done and how you landed in the situation ? As explained above stating "me too" is not helpful at all. You are not helping yourself, you  are not not helping others. you give no-one a chance to help you.

**Reply by iamhimanshu0** [reply]
> It does not give you anything if we do.what do you think would happen if we do?.
> 
> There is no reproducible csse to be able to diagnose it. Unless you help with making a reproducible case, there is no way anyone can help, i am afraid.

can u help me make a reproducible case? i did tried everything but result are the same..

**Reply by potiuk** [reply]
You are the only one who can help yourself. There is no way anyone can help if you don't describe what you have done to get there. 

Be aware - that people here help in their free time. There is a great documentation for starting your journey with airflow, people succeeded multiple times following it.

Start here: https://airflow.apache.org/docs/apache-airflow/stable/start/index.html but there is a lot of useful information to follow from there. 

If you did not succeded, you must have done something different than others. DON't. Check prerequisites. Make sure they are followed. Read the docs and try to understand the docs before making each step. If you do not understand something described there, read up more about the subject you don't understand.

If in doubt - start from scratch and - again - follow the installation documentation **precisely**.  That includes setting the configuration, creating the databases and so on.

**Reply by potiuk** [reply]
And finally - if you did not succeed, while precisely following the docs, describe in detail what you have done, what operating system you have, what steps you have taken, what errors you received. But do not show the final outcome. Show the journey how you get there. Then **maybe** someone will be able to help.

**Reply by yenchenLiu** [reply]
Hi, I also face the same issue, and I have used airflow for a year.

I just upgrade the airflow version from `2.2.3` to `2.2.4`, and I realize that the airflow adds the session backend mechanism.

However, the problem is that I have already migrated the DB(SQLite for development), and I checked the DB which also has a `session` table. I feel confused.
 
I guess `flask` may not find the DB correctly. I will figure it out tomorrow. If I find something new, I will reply here.

**Reply by potiuk** [reply]
Ah. That might ring a bell then @uranusjr @ephraimbuddy @jhtimmins @kaxil @ashb - did not we have a problem somewhere that the schema for FAB data was wrongly set? 

I cannot find it easily but it does ring a bell.

**Reply by yenchenLiu** [reply]
I found the problem, and `Flask` doesn't get the correct configuration.

I tried to log the path of the DB file in flask-session, and I got `<SQLAlchemy engine=sqlite://[virtual-environment]/lib/python3.9/site-packages/airflow/www/airflow.db>
`

My DB connection string is `sqlite:///airflow.db` which is a relative path. When I use an absolute path such as `sqlite:////tmp/airflow.db`, it works.

When setting the relative path, Flask considers the relative path to come from `airflow/www`

Thank you.

**Reply by potiuk** [reply]
AHA. That's an interesting one.

Sounds like a good candidate to make PR of (checking if the sqlalchemy URl is relative and raising an error in this case @yenchenLiu  - would you like maybe to contribute such PR ? Would ba a nice contribution and would prevent people from making the same mistake!

**Reply by yenchenLiu** [reply]
OK.
I opened a PR, please let me know if there are any problems.
https://github.com/apache/airflow/pull/22530

**Reply by potiuk** [reply]
Thanks a lot for being dilligent about it. That was hell-of-an issue without a clue where it came from. Sorry for being a bit harsh initially (or it could look like that). It's really cool @yenchenLiu found it!  

We are going to cherry-pick #22530 when the tests pass for 2.2.5 to prevent similiar issues. And 2.2.5 is going to be released today or tomorrow (or Monday). @yenchenLiu that might make you one of the fastest road from "new contributor" to "code released" in Airflow :)

**Reply by iamhimanshu0** [reply]
> 

Thank you so much. it's working fine (*)

**Reply by nathan-downs** [reply]
I ran into this same issue, but was able to solve it by running `airflow db reset` to get the session table created because `airflow db init` was not working for me (no errors, but something was silently failing).

**Comment by puranjay123** [comment]
I encountered this error while setting up Airflow and based on above answers I think I should summarize it more clearly if you face this issue
`(sqlite3.OperationalError) no such table: session `
then 
### 1st answer by @yenchenLiu 
Go to `airflow.cfg` file<br>
and change this  to your absolute path like this

`sql_alchemy_conn = sqlite:////home/puranjay/Desktop/Apache Airflow/airflow.db
`
Ideally it should update this location itself but if it did't. Change the path to your airflow.db absolute path

### 2nd answer  by @nathan-downs which worked for me
the session tables are not created which is caused when creating the db there might be a issue which causes tables to not create properly so  do just 
`airflow db reset` 
this should work properly.
then create or up the webserver again everything should work fine now.
